### PR TITLE
Annotate how query works

### PR DESCRIPTION
**Pull request by MorrisJobke**
https://github.com/phan/phan/wiki/Generic-Types#indicating-a-method-creates-an-instance-of-the-passed-in-class-name

Extracted from #18680 by @kesselb

### CODE REVIEW

**Comment by kesselb** [comment]
Yep. Should also work with Psalm and hopefully PHPStorm sometime.

**Comment by MorrisJobke** [comment]
> Yep. Should also work with Psalm and hopefully PHPStorm sometime.

Does work, but now Psalm stumbles over the non-class keys. :/ Anyways - something that we can look into with this separate PR.

**Comment by MorrisJobke** [comment]
> Anyways - something that we can look into with this separate PR.

Things that are good to clean up :)

**Comment by kesselb** [comment]
Ah. At some places we are still using ->query('UserStatus'). Psalm is unable to resolve those aliases and fails.

**Comment by MorrisJobke** [comment]
> Ah. At some places we are still using ->query('UserStatus'). Psalm is unable to resolve those aliases and fails.

Exactly - but this is something we can tackle in 21.

**Review by rullzer — APPROVED** [review]
:see_no_evil: sure

File: lib/private/AppFramework/Utility/SimpleContainer.php [hunk 1]
```diff
@@ -52,6 +52,12 @@ public function __construct() {
 		$this->container = new Container();
 	}
 
+	/**
+	 * @template T
+	 * @param class-string<T> $id
```
Comment (ChristophWurst) [inline lib/private/AppFramework/Utility/SimpleContainer.php:57]: what if you use `class-string<T>|string` for now? would that work with our named aliases?

**Comment by MorrisJobke** [comment]
Implemented via #23377

**Pull request closed without merging**